### PR TITLE
guix: init at 1.1.0

### PR DESCRIPTION
**Pull request by bqv**
###### Motivation for this change
Add Guix (migrated from #84004 because I'm a moron)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bqv** [comment]
If you'd like to actually test this, I have a module-ish sufficient [here](https://github.com/bqv/nixos/raw/live/profiles/misc/guix.nix) that allows me to `guix environment -C guix --ad-hoc bash vim guile-json` (which fetched the relevant packages, grafted a few, built a profile and dropped me into a working shell)

**Comment by Emiller88** [comment]
Just for future reference what does this have over https://github.com/NixOS/nixpkgs/pull/56430?

**Comment by bqv** [comment]
This is just the package without the module for now, builds a functional guix from source rather than using the guix binary bootstrap package, and isn't a year-old PR :p

**Comment by cole-h** [comment]
@ofborg build

**Review by cole-h — APPROVED** [review]
Diff LGTM, `guix --help` displays a long list of subcommands, and borg is happy.

```
[16 built, 18 copied (53.8 MiB), 11.7 MiB DL]
https://github.com/NixOS/nixpkgs/pull/85463
6 packages built:
guile-gcrypt guile-git guile-json guile-sqlite3 guile-ssh guix
```

File: pkgs/development/guile-modules/guile-gcrypt/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  version = "0.2.1";
+
+  src = fetchurl {
+    url = "https://notabug.org/cwebber/${pname}/archive/v${version}.tar.gz";
+    sha256 = "1qj1yw0kman984x584jjjxnjdhm0pwgp09iyn3b5rqajx7svpqcd";
+  };
+
+  postConfigure = ''
+    sed -i '/moddir\s*=/s%=.*%=''${out}/share/guile/site%' Makefile;
+    sed -i '/godir\s*=/s%=.*%=''${out}/share/guile/ccache%' Makefile;
```
Comment (timokau) [inline pkgs/development/guile-modules/guile-gcrypt/default.nix:16]: Is it not possible to set these variables through environment variables or pass them to make somehow?

File: pkgs/development/guile-modules/guile-gcrypt/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  nativeBuildInputs = [ autoreconfHook pkgconfig texinfo ];
+  buildInputs = [ guile ];
+  propagatedBuildInputs = [ libgcrypt ];
+
+  meta = with stdenv.lib; {
+    description = "Bindings to Libgcrypt for GNU Guile";
+    homepage = "https://notabug.org/cwebber/guile-gcrypt";
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ bqv ];
```
Comment (timokau) [inline pkgs/development/guile-modules/guile-gcrypt/default.nix:27]: The commit order doesn't seem right. `bqv` is not defined yet here. The maintainer commit should be the first one.

File: pkgs/development/guile-modules/guile-git/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
+{ stdenv, fetchFromGitHub, fetchFromGitLab, guile, libgit2
+, autoreconfHook, pkg-config, texinfo
+}:
+
+let
+  bytestructures = stdenv.mkDerivation rec {
```
Comment (timokau) [inline pkgs/development/guile-modules/guile-git/default.nix:6]: Why did you define this inline and not in its own package?

File: pkgs/development/guile-modules/guile-git/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    src = fetchFromGitHub {
+      owner = "TaylanUB";
+      repo = pname;
+      rev = "v${version}";
+      sha256 = "0q0habjiy3h9cigb7q1br9kz6z212dn2ab31f6dgd3rrmsfn5rvb";
+    };
+
+    postConfigure = ''
+      sed -i '/moddir\s*=/s%=.*%=''${out}/share/guile/site%' Makefile;
+      sed -i '/godir\s*=/s%=.*%=''${out}/share/guile/ccache%' Makefile;
```
Comment (timokau) [inline pkgs/development/guile-modules/guile-git/default.nix:19]: Same comment from before applies here. If we need to do this multiple times, that's even more reason to find a better way.

File: pkgs/development/guile-modules/guile-git/default.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    nativeBuildInputs = [ autoreconfHook pkg-config ];
+    buildInputs = [ guile ];
+  };
+in stdenv.mkDerivation rec {
+  pname = "guile-git";
+  version = "0.3.0";
+
+  src = fetchFromGitLab {
+    owner = pname;
```
Comment (timokau) [inline pkgs/development/guile-modules/guile-git/default.nix:30]:
```suggestion
    owner = "guile-git";
```

The owner happens to match `pname`, but the two values don't change together. Preferably `repo` should be explicit as well.

File: pkgs/development/guile-modules/guile-ssh/default.nix [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    owner = "artyom-poptsov";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "054hd9rzfhb48gc1hw3rphhp0cnnd4bs5qmidy5ygsyvy9ravlad";
+  };
+
+  configureFlags = [ "--with-guilesitedir=\${out}/share/guile/site" ];
+
+  postFixup = ''
+    for f in $out/share/guile/site/ssh/**.scm; do \
```
Comment (timokau) [inline pkgs/development/guile-modules/guile-ssh/default.nix:19]: Its somewhat hard to believe that something used in `guix` is not using fully-qualified paths. Why is this necessary?

File: pkgs/tools/package-management/guix/default.nix [hunk 7]
```diff
@@ -0,0 +1,81 @@
+{ stdenv, pkgs, lib, fetchurl, pkg-config, makeWrapper, zlib, bzip2, guile
+, guile-gcrypt, guile-git, guile-json, guile-sqlite3, guile-ssh, gnutls
+, storeDir ? null, stateDir ? null }:
+
+let
+  guile-gnutls = (gnutls.override {
+    inherit guile;
+    guileBindings = true;
+  }).overrideAttrs (attrs: {
+    # Workaround until https://github.com/NixOS/nixpkgs/pull/80206 is allowed in
```
Comment (timokau) [inline pkgs/tools/package-management/guix/default.nix:10]: Lets get #80206 done first instead of adding workarounds here.

File: pkgs/tools/package-management/guix/default.nix [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    guile-gcrypt guile-git guile-json guile-sqlite3 guile-ssh guile-gnutls.out guile-git.bytestructures
+  ];
+
+  nativeBuildInputs = [ pkg-config makeWrapper ];
+  buildInputs = [ zlib bzip2 ] ++ modules;
+  propagatedBuildInputs = [ guile ];
+
+  GUILE_LOAD_PATH = let
+    guilePath = [
+      "\${out}/share/guile/site"
```
Comment (timokau) [inline pkgs/tools/package-management/guix/default.nix:41]: Won't this try to reference `$out` at runtime?

File: pkgs/tools/package-management/guix/default.nix [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      --prefix GUILE_LOAD_PATH : "${GUILE_LOAD_PATH}" \
+      --prefix GUILE_LOAD_COMPILED_PATH : "${GUILE_LOAD_COMPILED_PATH}"
+  '';
+
+  passthru = {
+    inherit guile;
+  };
+
+  meta = with lib; {
+    description = "Functional package manager for installed software packages and versions";
```
Comment (timokau) [inline pkgs/tools/package-management/guix/default.nix:75]: Is that an official tagline? The "installed" part doesn't make much sense to me.

**Review by timokau — CHANGES_REQUESTED** [review]
Excited to see this merged!

File: pkgs/development/guile-modules/guile-gcrypt/default.nix (hunk 1 above)
Comment (bqv) [reply pkgs/development/guile-modules/guile-gcrypt/default.nix:16]: No (or I would have done :p)

I know this is flimsy, but it's also a very short Makefile so there's little room for error here.

File: pkgs/development/guile-modules/guile-gcrypt/default.nix (hunk 2 above)
Comment (bqv) [reply pkgs/development/guile-modules/guile-gcrypt/default.nix:27]: It is actually the first one - presumably you're looking at commit dates or something similar which will be very very wrong by now due to my "creative" rebases

File: pkgs/development/guile-modules/guile-git/default.nix (hunk 3 above)
Comment (bqv) [reply pkgs/development/guile-modules/guile-git/default.nix:6]: This looks to have been some confusion on my part, I thought it was implemented in a different lisp and therefore not a guile module, but I've just checked and it is...

File: pkgs/development/guile-modules/guile-git/default.nix (hunk 4 above)
Comment (bqv) [reply pkgs/development/guile-modules/guile-git/default.nix:19]: I'm open to ideas, but it's hard coded in the Makefile. I don't see any other way beyond text manipulation

File: pkgs/tools/package-management/guix/default.nix (hunk 8 above)
Comment (bqv) [reply pkgs/tools/package-management/guix/default.nix:41]: No, it falls through to be evaluated while creating the wrappers in installPhase. There may be a better way of quoting them, though, I'm not sure

File: pkgs/tools/package-management/guix/default.nix (hunk 9 above)
Comment (bqv) [reply pkgs/tools/package-management/guix/default.nix:75]: It was, it's since changed...

File: pkgs/development/guile-modules/guile-ssh/default.nix (hunk 6 above)
Comment (bqv) [reply pkgs/development/guile-modules/guile-ssh/default.nix:19]: See https://github.com/guix-mirror/guix/blob/35c43fcdbb408a5755efebc13241dd9082360aa1/gnu/packages/ssh.scm#L275-L277

File: pkgs/development/guile-modules/guile-git/default.nix (hunk 3 above)
Comment (bqv) [reply pkgs/development/guile-modules/guile-git/default.nix]: Migrated to it's own file

File: pkgs/tools/package-management/guix/default.nix (hunk 9 above)
Comment (bqv) [reply pkgs/tools/package-management/guix/default.nix]: Updated to something more sensical :)

File: pkgs/development/guile-modules/bytestructures/default.nix [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+  src = fetchFromGitHub {
+    owner = "TaylanUB";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "0q0habjiy3h9cigb7q1br9kz6z212dn2ab31f6dgd3rrmsfn5rvb";
+  };
+
+  postConfigure = ''
+    sed -i '/moddir\s*=/s%=.*%=''${out}/share/guile/site%' Makefile;
```
Comment (doronbehar) [inline pkgs/development/guile-modules/bytestructures/default.nix:17]: Did you try something like this:

https://github.com/NixOS/nixpkgs/blob/fa0849d02308ea4d0b646f6fffff07cd042ab810/pkgs/development/guile-modules/guile-ncurses/default.nix#L17-L19

Or perhaps even better:

```nix
configureFlags = [
  "--with-guilesitedir=${placeholder "out"}/share/guile/site"
];

Comment (doronbehar) [reply pkgs/development/guile-modules/bytestructures/default.nix:17]: If that doesn't work I'd advice contacting upstream.

Comment (bqv) [reply pkgs/development/guile-modules/bytestructures/default.nix:17]: Yes, like I said to timo above, it's not something that can be substituted by autotools, it's hardcoded in the Makefile.

Comment (doronbehar) [reply pkgs/development/guile-modules/bytestructures/default.nix:17]: And what's wrong with this output:

```
$ tree $(nix-build -A guile-json)
result
├── lib
│   └── guile
│       └── 2.2
│           └── site-ccache
│               ├── json
│               │   ├── builder.go
│               │   └── parser.go
│               └── json.go
└── share
    └── guile
        └── site
            └── 2.2
                ├── json
                │   ├── builder.scm
                │   └── parser.scm
                └── json.scm

10 directories, 6 files
```

Comment (doronbehar) [reply pkgs/development/guile-modules/bytestructures/default.nix:17]: This is what you get if you don't specify any configure flags nor apply that postConfigure hook.

Comment (bqv) [reply pkgs/development/guile-modules/bytestructures/default.nix:17]: These paths have to be linked at by guile, for every package needed by guix. Note that some of these packages by default do not have 2.2 in that path, some don't even properly specify their paths (and therefore end up trying to write directly to the guile derivation), or have extensions and try and write those somewhere that they won't be found. For the sake of consistency across guile modules and sanity of everyone trying to build this, I thought it might be nice to fix those

Comment (doronbehar) [reply pkgs/development/guile-modules/bytestructures/default.nix:17]: > These paths have to be linked at by guile

Are you talking about the wrapping done at `guix/default.nix` that adds `GUILE_LOAD_COMPILED_PATH` to the environment?

It just seems weird to me, that a package maintained by Guix's developers (at least in a certain sense) - a package that's needed to bootstrap a package manager which is based on Nix, will need such hacky workarounds to fix paths in `$out`. Guix doesn't put these to any of those guile dependencies, see:

- https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/guile.scm?id=5d9e2187929ed7e8d46ec3cb3174fd78c1846360#n490
- https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/guile.scm?id=5d9e2187929ed7e8d46ec3cb3174fd78c1846360#n654

Moreover, they have Guile 3.0.

**Comment by bqv** [comment]
This is exhausting. I'm moving this to my local repository.

I'm going to close this because I'm not a fan of having ancient PRs open on my account which seem to have no hope of being merged.

For future reference, if anyone would like to use this, as of this head it is quite functional and I'm actually using it on my local machine. It is implemented in the sanest way I could think of, but if anyone thinks they can do better then please, *please* do.

File: pkgs/tools/package-management/guix/default.nix (hunk 8 above)
Comment (timokau) [reply pkgs/tools/package-management/guix/default.nix:40]: Maybe `placeholder "out"` would be better. At least that way its more obvious.

**Comment by timokau** [comment]
Sad to hear that, thank you for your work so far. If someone wants to pick this up where @bqv left off, feel free to ping me in the new PR if you want a review.
